jormungandr/src/blockchain/tip.rs

```diff
@@ -1,12 +1,214 @@
+            .ok_or(storage::Error::BlockNotFound)?;
+
+        match chain_selection::compare_against(storage, &tip_ref, &candidate) {
+            ComparisonResult::PreferCurrent => {
+                tracing::info!(
+                    "create new branch with tip {} | current-tip {}",
+                    candidate.header().description(),
+                    tip_ref.header().description(),
+                );
+                self.blockchain
+                    .branches_mut()
+                    .apply_or_create(candidate.clone())
+                    .await;
+            }
+            ComparisonResult::PreferCandidate => {
+                if tip_ref.hash() == candidate.block_parent_hash() {
+                    tracing::info!(
+                        "update current branch tip: {} -> {}",
+                        tip_ref.header().description(),
+                        candidate.header().description(),
+                    );
+
+                    storage.put_tag(MAIN_BRANCH_TAG, candidate_hash)?;
+
+                    let fragment_ids = block.fragments().map(|f| f.id()).collect();
+                    self.try_request_fragment_removal(fragment_ids, &block.header())?;
+
+                    self.tip.update_ref(candidate.clone()).await;
+                } else {
+                    let common_ancestor =
+                        storage.find_common_ancestor(candidate_hash, tip_ref.hash())?;
+
+                    let stream = self
+                        .blockchain
+                        .storage()
+                        .stream_from_to(common_ancestor, candidate_hash)?;
+                    tokio::pin!(stream);
+
+                    // there is always at least one block in the stream
+                    let ancestor = stream.next().await.unwrap()?;
+                    if let Some(ref mut mbox) = self.fragment_mbox {
+                        mbox.try_send(TransactionMsg::BranchSwitch(ancestor.date().into()))?;
+                    }
+
+                    while let Some(block) = stream.next().await {
+                        let block = block?;
+                        let fragment_ids = block.fragments().map(|f| f.id()).collect();
+                        self.try_request_fragment_removal(fragment_ids, &block.header())?;
+                    }
+
+                    tracing::info!(
+                        "switching branch from {} to {}",
+                        tip_ref.header().description(),
+                        candidate.header().description(),
+                    );
+
+                    self.blockchain
+                        .storage()
+                        .put_tag(MAIN_BRANCH_TAG, candidate_hash)?;
+
+                    let branch = self
+                        .blockchain
+                        .branches_mut()
+                        .apply_or_create(candidate.clone())
+                        .await;
+                    self.tip.swap(branch).await;
+                }
```

Comment:
This block is too long, could be refactored into smaller functions for readability.